IN-456: Add Github Action for applying Drupal 7 security updates

davialexandre opened the pull request:
## Overview

This adds the first version of the Github Action to automatically apply security updates to Drupal 7 sites. It only updates Drupal core and modules.

Please check the README file for more details on how it works and how to use it.

For an example of the action being used, check [this](https://github.com/compucorp/outdated-d7-site/runs/887980747?check_suite_focus=true). This is a sample PR created by the action: https://github.com/compucorp/outdated-d7-site/pull/22.

(Note: This `outdated-d7-site` is just a sample project created to demonstrate the action. It will be deleted in the near future).

## Technical details
The main part of the action is inside the entrypoint of the docker image (the `entrypoint.sh` file). However, there are some steps in the update process that would not be possible to be easily implemented with Bash scripts (generating a markdown message for the commit/Pull Request and creating the Pull Request using the Github API). Due to that, the action also depends on two small PHP scripts where it was much easier to implement these more complicated pieces of logic.

This PR also adds a workflow to run linters for Dockerfiles (`hadolint`), Bash Scripts (`shellcheck`) and PHP (`phpcs`). For PHP, since this is not a Drupal, CiviCRM or Symfony project, the code standard being followed is [PSR-12](https://www.php-fig.org/psr/psr-12/).

davialexandre committed:
IN-456: Add Github Action for applying Drupal 7 security updates

reneolivo commented on README.md:
````diff
@@ -0,0 +1,91 @@
+```
+
+Note that in order to be able to access the output of a step, you need to give it an `id` (`security-updates` in this 
+case). Also note that `branch` will be empty in case no security updates were applied. We can use that inside and `if` 
````
We can use that inside **an** `if`

reneolivo commented on .github/workflows/linter.yml:
```diff
@@ -0,0 +1,22 @@
+        run : hadolint --config .hadolint.yml Dockerfile
+
+      - name: Run shellcheck
+        uses: ludeeus/action-shellcheck@0.4.1
```
There is no mention of this action in the PR description. Why do we need it? Can you add it to the description? I'm assuming it checks the Dockerfile configurations.

davialexandre replied:
It's a linter for bash scripts. I'll make that clear

reneolivo commented on build-update-message.php:
```diff
@@ -0,0 +1,43 @@
+ */
+$handle = fopen('php://stdin', 'rb');
+
+if ($handle === false) {
```
`FALSE`

davialexandre replied:
I decided to follow PSR-12 as the code standard for this project, so `false` is actually valid.

reneolivo commented on build-update-message.php:
```diff
@@ -0,0 +1,43 @@
+while (($data = fgetcsv($handle, 1000, ",")) !== FALSE) {
+    // Basic validation.
+    // If the input doesn't contain 4 fields, we just stop processing it
+    if (count($data) != 4) {
```
!==

reneolivo commented on create-pull-request.php:
```diff
@@ -0,0 +1,86 @@
+curl_setopt($ch, CURLOPT_USERAGENT, 'Drupal7-Security-Updates');
+curl_setopt($ch, CURLOPT_USERPWD, "$user:$token");
+curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode(
+  [
```
are we indenting by 4 spaces or by 2? everywhere else is 4 spaces.

davialexandre replied:
4 spaces, according to PSR-12

reneolivo commented on create-pull-request.php:
```diff
@@ -0,0 +1,86 @@
+            throw new RuntimeException('Error creating pull request: ' . $error->message);
+        }
+    }
+}
```
Why create a whole script for this when there's an action that can be used?
https://github.com/marketplace/actions/create-pull-request

Or, is it that the action can't be used?

davialexandre replied:
Mainly for 2 reasons:
- Security: Internally, any action that creates a Pull Request will need a token that gives write access to the repo. I don't like the idea of giving access to that code to some code written by someone else, even though it will be a temporary token created for a single run of the workflow
- Simplicity: We'll need add security update workflows to a dozen different repos, to the fewer steps we have in the workflow, the less we'll need to duplicate

reneolivo replied:
Ummmm, you know your OS was written by someone else, right? 😆 

It's ok to include this script. The only question I have (as a side note) is: how are we going to import the action into our own repos? Copying it or using the marketplace?

davialexandre replied:
> how are we going to import the action into our own repos? Copying it or using the marketplace?

Neither. It's just going to be used an regular action. See [here](https://github.com/compucorp/outdated-d7-site/blob/master/.github/workflows/security-updates.yml#L28) for an example (Using `@master` there because we don't have a tag yet)

reneolivo commented on create-pull-request.php:
```diff
@@ -0,0 +1,86 @@
+curl_setopt($ch, CURLOPT_USERPWD, "$user:$token");
+curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode(
+  [
+    'title' => 'Automated Security Updates',
```
Can we add the tag name to the title? 🐶

davialexandre replied:
I thought a bit about this and I think it's not worth it. In order to be able to do that, I'd have to either pass an additional param to the script or extract it from the branch name. Besides this, I kept thinking on what difference it would really make to have the tag in title and I couldn't think of any. In the end, the important thing is to know that we have security updates available and that they should be merged as quickly as possible.

Anyway, I have followed your suggestion to add the tag number to the note in the Pull Request, so I guess that would be a good enough way for people to know which tag was used?

reneolivo commented on entrypoint.sh:
```diff
@@ -0,0 +1,71 @@
+echo "::set-output name=branch::$security_updates_branch"
+
+# Let people know how the branch was created and how to fix conflicts, if there are any
+[ $base_branch_or_tag != "master" ] && printf "\n\nImportant note: This update was created from a tag. Conflicts are expected in case \`master\` is ahead the tag. If that is the case, the conflicts should be manually fixed in a new branch" >> /update-message.txt
```
Should we also include the tag that was used in this message? I know the branch name has it, but this is more for humans.

davialexandre replied:
Done

reneolivo reviewed (changes requested):
@davialexandre thanks for this. The changes look great. I think my only major problem is the use of a custom script for creating a PR when there's already an action for it.

davialexandre committed:
IN-456: Add PHP linter

Also, in order to make it easier to add the PHP linter, the way the workflow works changed a little bit.

Now, instead running it inside the hadolint container, we run things on top of ubuntu which already comes installed with most of the things we need.

The Coder standard being followed is PSR-12. Since this is not a CiviCRM, Drupal or Symfony project, it doesn't make sense to follow any of their code standards, so PSR-12 was chosen for being the most generic and recent one available.

reneolivo commented on create-pull-request.php:
```diff
@@ -55,31 +59,30 @@
  *
- * It's not easy to check if a PR for a branch already exists, so, instead, we
- * simply always try to create one and ignore the error.
+ * It's not easy to check if a PR for a branch already exists, so, instead, we imply always
```
*imply* typo?

davialexandre replied:
fixed
